EventFilter/EcalRawToDigi/plugins/alpaka/UnpackPortable.dev.cc

```diff
@@ -0,0 +1,441 @@
+#include <alpaka/alpaka.hpp>
+
+#include "DataFormats/DetId/interface/DetId.h"
+#include "DataFormats/EcalDigi/interface/EcalConstants.h"
+#include "EventFilter/EcalRawToDigi/interface/ElectronicsIdGPU.h"
+#include "EventFilter/EcalRawToDigi/interface/DCCRawDataDefinitions.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/workdivision.h"
+
+#include "UnpackPortable.h"
+
+namespace ALPAKA_ACCELERATOR_NAMESPACE::ecal::raw {
+
+  using namespace ::ecal::raw;
+  using namespace cms::alpakatools;
+
+  class Kernel_unpack {
+  public:
+    template <typename TAcc, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+    ALPAKA_FN_ACC void operator()(TAcc const& acc,
```

Comment:
There are two potential issues with this kernel:
  - it _must_ be called with the correct number of blocks;
  - it ignores the _elements_ per thread.

However, after re-reading the implementation, I think the implementation is correct, and these issues affect at most the potential for optimisations. 

The first issue is not problematic, as long as the call to `Kernel_unpack` (on lines [428..439](https://github.com/cms-sw/cmssw/pull/43257/files#diff-58b34a0c469e4fb6517d2f679da7cd8569321e1d695b6fd1085f2bf2bb354ff1R428-R439)) uses the correct number of blocks.
A more flexible approach would be to pass the number of FEDs as an additional argument, and add an outer loop so that a smaller number of blocks could process more FEDs.
However, this is only a potential optimisation that would allow tuning the number of blocks, and an be left to a later time - if this kernel is ever identified as slow.

Regarding the second issue, the current approach works because each "block" does anyway an internal loop over the channels.
On the CPU the kernel launch does set the number of elements per thread to 32, but this is ignored.
However, taking it into account would result in two nested loops, but effectively process the data in the same way.

Reply:
I would leave this for a later round. There will be some changes needed to the unpacker anyway when adding the unpacking of the auxiliary data and in terms of computation time the kernel is not very heavy anyway as it is now.

Reply:
Agreed.

Do we already have an issue on GitHub to keep track of follow up improvements for the ECAL Alpaka PRs ?

If not, could you create one ?

Reply:
Yes we do. You created it :-)
https://github.com/cms-sw/cmssw/issues/43551

Reply:
OK, I've copied these comments there.